core/jvm/src/main/scala/fs2/interop/flow/StreamSubscriber.scala

```
@@ -0,0 +1,256 @@
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2
+package interop
+package flow
+
+import cats.MonadThrow
+import cats.effect.kernel.Async
+import cats.syntax.all._
+
+import java.util.concurrent.Flow.{Subscriber, Subscription}
+import java.util.concurrent.atomic.AtomicReference
+
+/** Implementation of a [[Subscriber]].
+  *
+  * This is used to obtain a [[Stream]] from an upstream reactive-streams system.
+  *
+  * @see [[https://github.com/reactive-streams/reactive-streams-jvm#2-subscriber-code]]
+  */
+private[flow] final class StreamSubscriber[F[_], A](
+    private[flow] val subscriber: StreamSubscriber.FSM[F, A]
+)(implicit
+    F: MonadThrow[F]
+) extends Subscriber[A] {
+
+  /** Called by an upstream reactive-streams system. */
+  override def onSubscribe(subscription: Subscription): Unit = {
+    nonNull(subscription)
+    subscriber.onSubscribe(subscription)
+  }
+
+  /** Called by an upstream reactive-streams system. */
+  override def onNext(a: A): Unit = {
+    nonNull(a)
+    subscriber.onNext(a)
+  }
+
+  /** Called by an upstream reactive-streams system. */
+  override def onComplete(): Unit =
+    subscriber.onComplete()
+
+  /** Called by an upstream reactive-streams system. */
+  override def onError(t: Throwable): Unit = {
+    nonNull(t)
+    subscriber.onError(t)
+  }
+
+  /** Creates a [[Stream]] from this [[Subscriber]]. */
+  def stream(subscribe: F[Unit]): Stream[F, A] =
+    subscriber.stream(subscribe)
+}
+
+private[flow] object StreamSubscriber {
+
+  /** Instantiates a new [[StreamSubscriber]] for the given buffer size. */
+  def apply[F[_], A](bufferSize: Int)(implicit F: Async[F]): F[StreamSubscriber[F, A]] =
+    fsm[F, A](bufferSize).map(fsm => new StreamSubscriber(subscriber = fsm))
```

Should we throw an exception if `bufferSize` is not positive?

---

Sure, I think Cats Effect does this in `std` with `require`. So this value can't be zero?

---

Luis explained offline that the `bufferSize` is used for `request(...)`, which is required positive by the spec.